src/modules/Invoice/pdf_template/default-pdf.twig

```diff
@@ -0,0 +1,78 @@
+{% if html_buyer_lines >= html_seller_lines %}
+	{% set top = 325 + (25 * html_buyer_lines) %}
+{% else %}
+	{% set top = 325 + (25 * html_seller_lines) %}
+{% endif %}
+
+<!DOCTYPE html>
+<html>
+	<head>
+		<meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
+		<title>{{ invoice.serie_nr }}</title>
+		<style>
+			{{css}}</style>
+	</head>
+	<body>
+		{{ html_logo|raw }}
+		<hr class='Rounded'>
+		<div class='InvoiceInfo'>
+			<p>Invoice number:
+				{{ invoice.serie_nr }}</p>
+			<p>Invoice date:
+				{{ invoice.created_at|fossbilling_format_time }}</p>
+			<p>Due date:
+				{{ invoice.due_at|fossbilling_format_time }}</p>
+			<p>Invoice status:
+				{{ invoice.status|capitalize }}</p>
+		</div>
+
+		<h3 class='CompanyInfo'>Company</h3>
+		<div class='CompanyInfo'>{{ html_seller|raw }}</div>
```

Comment:
I'd rather if we removed an empty / null data and then passed the data along to twig rather than creating the list from within PHP. Doing it that way would be more flexible and remove the need for the `|raw`, which can help prevent against embedded HTML being added to the info

Reply:
Could you explain further? I understand that you want to get rid of the `|raw` filter. But that would mean to  "translate" the PHP functions `getSellerPdfHtml` and `getBuyerPdfHtml` to twig. Am I seeing that correctly?

Reply:
More-or-less, yes. Ultimately, the intention is to not have that hard-coded so that custom PDF templates may draw the list however they like and additionally remove the usage of the `raw` filter.

However it isn't *super* important how it's done as long as we enure that the PDF won't for example have an empty spot for a client's company if they aren't part of a company. In theory, this can be done twig-side by using the [length](https://twig.symfony.com/doc/3.x/filters/length.html) filter as the data should always be set and we just care that it has an actual length to it.

A solution that's 50/50 on the PHP & twig usage do something like this and then pass the array off to twig:

```PHP
$sourceData = [
    'Name' => $invoice['seller']['company'],
    'Address' => $invoice['seller']['address'],
    'Company Vat' => $invoice['seller']['company_vat'],
    'Company Number' => $invoice['seller']['company_number'],
    'Phone' => $invoice['seller']['phone'],
    'Email' => $invoice['seller']['email'],
];

foreach ($sourceData as $label => $data) {
    if(empty($data)){
        unset($sourceData[$label]);
    }
}
```